crates/burn-jit/src/kernel/conv/conv2d/gemm/homogeneous/base.rs

```diff
@@ -0,0 +1,450 @@
+        },
+        kernels::matmul::AdvancedConfig,
+    },
+    prelude::*,
+};
+use std::marker::PhantomData;
+
+use crate::kernel::conv::{
+    conv2d::gemm::base::{Convolution, ConvolutionKernel, ConvolutionLaunch, ConvolutionProblem},
+    loader::im2col::SimpleIm2colLoader,
+};
+use crate::kernel::conv::{
+    conv2d::gemm::Config as _,
+    loader::{bias::BiasLoader, Loader},
+};
+
+/// Performs matrix multiplication at the global level, with each plane sharing the same responsibilities
+/// - All planes load data to the stage
+/// - All planes are used in the stage matmul computation
+pub struct ImplicitGemmConvolution<
+    EG: Numeric,
+    ES: Numeric,
+    Acc: Numeric,
+    SMM: stage::Matmul<ES, EG, Acc>,
+> {
+    _eg: PhantomData<EG>,
+    _es: PhantomData<ES>,
+    _acc: PhantomData<Acc>,
+    _stage_matmul: PhantomData<SMM>,
+}
+
+#[cube]
+impl<EG, ES, Acc, SMM, SMMConf> Convolution<EG, ES, Acc, SMM>
+    for ImplicitGemmConvolution<EG, ES, Acc, SMM>
+where
+    EG: Numeric,
+    ES: Numeric,
+    Acc: Numeric,
+    SMMConf: stage::Config,
+    SMM: stage::Matmul<
+        ES,
+        EG,
+        Acc,
+        LhsReader = LhsReader<ES>,
+        RhsReader = RhsReader<ES>,
+        Config = SMMConf,
+    >,
+{
+    type LhsLoader = SimpleIm2colLoader<EG, ES, Self::Config>;
+    type RhsLoader = RhsLoader<EG, ES, SMM::Config, CyclicLoading>;
+    type AccumulatorLoader = BiasLoader<EG, Acc, SMM::Config>;
+
+    type Out = Unloader<EG>;
+    type Accumulator = SMM::Accumulator;
+
+    fn execute(
+        mut lhs_loader: Self::LhsLoader,
+        mut rhs_loader: Self::RhsLoader,
+        mut acc_loader: Self::AccumulatorLoader,
+        mut out_unloader: Self::Out,
+        acc: &mut Self::Accumulator,
+        k_range: (u32, u32),
+        #[comptime] config: Self::Config,
+    ) {
+        let k_step = SMM::K;
+        let range = k_range.1 - k_range.0;
+        let num_loops = (range + k_step - 1) / k_step;
+
+        Self::AccumulatorLoader::fill_stage(&mut acc_loader, config.to_smm_config());
+        let (mut lhs_tile, mut rhs_tile) = SMM::init_tile_inputs(config.to_smm_config());
+
+        sync_units();
+
+        SMM::fill_accumulator::<Self::AccumulatorLoader>(
+            &mut acc_loader,
+            acc,
+            config.to_smm_config(),
+        );
+
+        for _ in 0..num_loops {
+            let lhs_stage_reader = &Self::LhsLoader::fill_stage(&mut lhs_loader, config);
+            let rhs_stage_reader = &Self::RhsLoader::fill_stage(&mut rhs_loader, config);
+
+            sync_units();
+
+            SMM::execute(
+                lhs_stage_reader,
+                rhs_stage_reader,
+                &mut lhs_tile,
+                &mut rhs_tile,
+                acc,
+                config.to_smm_config(),
+            );
+
+            sync_units();
+
+            Self::LhsLoader::advance_view(&mut lhs_loader, k_step);
+            Self::RhsLoader::advance_view(&mut rhs_loader, k_step);
+        }
+
```

Comment:
Somehow adding a `sync_units` after the for loop improved performance for the matmul. I think it makes sure all units in a plane are sync following the loop which improve the execution of following operations.

Reply:
I'll benchmark it

Reply:
So the benchmark results are very odd. I tried 4 different ways of syncing, and the one I used initially was overall the fastest for CUDA, but adding a sync before the load and after the loop was significantly faster for SPIR-V. Only syncing where absolutely needed was the slowest by far. Very odd behaviour. I'll stick with lots of sync for now because it's only 10-15% slower than the current implementation on CUDA (and there's margin of error), but 30% faster on SPIR-V.